lib/grandpa/message_handler_test.go

```diff
@@ -587,3 +587,251 @@ func TestMessageHandler_VerifyBlockJustification(t *testing.T) {
+		SetID:                gs.state.setID,
+		Round:                gs.state.round,
+		PreVoteJustification: prevotesJustification,
+		Hash:                 bfcHash,
+		Number:               uint32(bfcNumber),
+	}
+
+	hash, err := h.verifyPreVoteJustification(testCatchUpResponse)
+	require.NoError(t, err)
+	require.Equal(t, hash, bfcHash)
+}
+
+func Test_VerifyPreCommitJustification(t *testing.T) {
+	t.Parallel()
+
+	gs, st := newTestService(t)
+	h := NewMessageHandler(gs, st.Block)
+
+	// add 10 previous blocks
+	bfcBlock := addBlocksAndReturnTheLastOne(t, st.Block, 10)
+
+	bfcHash := bfcBlock.Header.Hash()
+	bfcNumber := bfcBlock.Header.Number.Int64()
+
+	// many of equivocatory votes
+	ed25519Keyring, _ := keystore.NewEd25519Keyring()
+	fakeAuthorities := make([]*ed25519.Keypair, 0)
+
+	// Alice, Charlie, David - Equivocatory
+	// Bob, Eve, Ferdie, Ian - Legit
+	// total of votes 4 legit + 3 equivocatory
+	// the threshold for testing is 9, so 2/3 of 9 = 6
+	fakeAuthorities = append(fakeAuthorities, ed25519Keyring.Alice().(*ed25519.Keypair))
+	fakeAuthorities = append(fakeAuthorities, ed25519Keyring.Alice().(*ed25519.Keypair))
+	fakeAuthorities = append(fakeAuthorities, ed25519Keyring.Alice().(*ed25519.Keypair))
+	fakeAuthorities = append(fakeAuthorities, ed25519Keyring.Bob().(*ed25519.Keypair))
+	fakeAuthorities = append(fakeAuthorities, ed25519Keyring.Charlie().(*ed25519.Keypair))
+	fakeAuthorities = append(fakeAuthorities, ed25519Keyring.Charlie().(*ed25519.Keypair))
+	fakeAuthorities = append(fakeAuthorities, ed25519Keyring.Charlie().(*ed25519.Keypair))
+	fakeAuthorities = append(fakeAuthorities, ed25519Keyring.Dave().(*ed25519.Keypair))
+	fakeAuthorities = append(fakeAuthorities, ed25519Keyring.Dave().(*ed25519.Keypair))
+	fakeAuthorities = append(fakeAuthorities, ed25519Keyring.Eve().(*ed25519.Keypair))
+	fakeAuthorities = append(fakeAuthorities, ed25519Keyring.Ferdie().(*ed25519.Keypair))
+	fakeAuthorities = append(fakeAuthorities, ed25519Keyring.Ian().(*ed25519.Keypair))
+
+	prevotesJustification := make([]SignedVote, len(fakeAuthorities))
+	for idx, fakeAuthority := range fakeAuthorities {
+		vote := types.GrandpaVote{
+			Hash:   bfcHash,
+			Number: uint32(bfcNumber),
+		}
+
+		sig := signFakeFullVote(
+			t, fakeAuthority, precommit, vote, gs.state.round, gs.state.setID)
+
+		prevotesJustification[idx] = SignedVote{
+			Vote:        vote,
+			Signature:   sig,
+			AuthorityID: fakeAuthority.Public().(*ed25519.PublicKey).AsBytes(),
+		}
+	}
+
+	testCatchUpResponse := &CatchUpResponse{
+		SetID:                  gs.state.setID,
+		Round:                  gs.state.round,
+		PreCommitJustification: prevotesJustification,
+		Hash:                   bfcHash,
+		Number:                 uint32(bfcNumber),
+	}
+
+	err := h.verifyPreCommitJustification(testCatchUpResponse)
+	require.NoError(t, err)
+}
+
+func signFakeFullVote(
+	t *testing.T, auth *ed25519.Keypair,
+	stage Subround, v types.GrandpaVote,
+	round, setID uint64) [64]byte {
+
```

Comment:
nit some linters (not ours yet :eyes:) will complain about the empty leading line :smile_cat: Let's remove it I'd say.

```suggestion
```